lib/codecs/src/encoding/format/cef.rs

```diff
@@ -0,0 +1,520 @@
+use crate::encoding::BuildError;
+use bytes::BytesMut;
+use chrono::SecondsFormat;
+use lookup::lookup_v2::ConfigTargetPath;
+use snafu::Snafu;
+use std::num::ParseIntError;
+use std::{collections::HashMap, fmt::Write};
+use tokio_util::codec::Encoder;
+use vector_core::{
+    config::DataType,
+    event::{Event, LogEvent, Value},
+    schema,
+};
+
+const DEFAULT_DEVICE_VENDOR: &str = "Datadog";
+const DEFAULT_DEVICE_PRODUCT: &str = "Vector";
+const DEFAULT_DEVICE_VERSION: &str = "0"; // Major version. TODO: find a way to get the actual vector version.
+const DEFAULT_EVENT_CLASS_ID: &str = "Telemetry Event";
+const DEVICE_VENDOR_MAX_LENGTH: usize = 63;
+const DEVICE_PRODUCT_MAX_LENGTH: usize = 63;
+const DEVICE_VERSION_MAX_LENGTH: usize = 31;
+const DEVICE_EVENT_CLASS_ID_MAX_LENGTH: usize = 1023;
+const NAME_MAX_LENGTH: usize = 512;
+const SEVERITY_MAX: u8 = 10;
+
+/// Represents the device settings in the CEF format.
+#[derive(Debug, Clone)]
+pub struct DeviceSettings {
+    pub vendor: String,
+    pub product: String,
+    pub version: String,
+    pub event_class_id: String,
+}
+
+impl DeviceSettings {
+    /// Creates a new `DeviceSettings`.
+    pub const fn new(
+        vendor: String,
+        product: String,
+        version: String,
+        event_class_id: String,
+    ) -> Self {
+        Self {
+            vendor,
+            product,
+            version,
+            event_class_id,
+        }
+    }
+}
+
+/// Errors that can occur during CEF serialization.
+#[derive(Debug, Snafu)]
+pub enum CefSerializerError {
+    #[snafu(display(
+        r#"LogEvent field "{}" exceed {} characters limit: actual {}"#,
+        field,
+        max_length,
+        actual_length
+    ))]
+    ExceededLength {
+        field: String,
+        max_length: usize,
+        actual_length: usize,
+    },
+    #[snafu(display(
+        r#"LogEvent CEF severity must be a number from 0 to {}: actual {}"#,
+        max_value,
+        actual_value
+    ))]
+    SeverityMaxValue { max_value: u8, actual_value: u8 },
+    #[snafu(display(r#"LogEvent CEF severity must be a number: {}"#, error))]
+    SeverityNumberType { error: ParseIntError },
+    #[snafu(display(r#"LogEvent extension keys can only contain ascii alphabetical characters: invalid key "{}""#, key))]
+    ExtensionNonASCIIKey { key: String },
+}
+
+/// Config used to build a `CefSerializer`.
+#[crate::configurable_component]
+#[derive(Debug, Clone)]
+pub struct CefSerializerConfig {
+    /// The CEF Serializer Options.
+    pub cef: CefSerializerOptions,
+}
+
+impl CefSerializerConfig {
+    /// Creates a new `CefSerializerConfig`.
+    pub const fn new(cef: CefSerializerOptions) -> Self {
+        Self { cef }
+    }
+
+    /// Build the `CefSerializer` from this configuration.
+    pub fn build(&self) -> Result<CefSerializer, BuildError> {
+        let device_vendor = escape_header(self.cef.device_vendor.as_str());
+        if device_vendor.len() > DEVICE_VENDOR_MAX_LENGTH {
```

Comment:
We can introduce a helper for this and the following checks:
```
fn escape_header(field: &str, max_length: usize, field_name: &'static str) -> Result<String, BuildError> {
        let escaped = escape_header(field);
        if escaped.len() > max_length {
            ExceededLengthSnafu {
                field: escaped.clone(),
                max_length,
                actual_length: escaped.len(),
            }
            .fail()?;
        }
        Ok(escaped)
    }
```

and then use it:
```
let device_vendor = validate_length(self.cef.device_vendor.as_str(), DEVICE_VENDOR_MAX_LENGTH, "device_vendor")?;
```

Reply:
good improvement, thanks